broadbean/segment.py

```diff
@@ -0,0 +1,111 @@
+# The module defining the segment object
+#
+#
+
+from typing import Callable, Dict, Union, Optional
+from inspect import signature
+
+import numpy as np
+
+# types for good hinting
+Number = Union[float, int]
+ArgsDict = Dict[str, Union[Number, str]]
+
+
+def validate_function_and_args_dict(func: Callable,
+                                    args_dict: ArgsDict) -> None:
+    """
+    Validate whether a function can be used in a segment
+    to generate an array and validate that the args_dict matches
+    the function.
+
+    The function must have the two arguments dur and SR and can
+    have as many other arguments as it will and the args_dict must
+    specify all these additional arguments
+    """
+    sig = signature(func)
+    argnames = [an for an in sig.parameters]
+
+    if 'dur' not in argnames:
+        raise ValueError('Function invalid, must have the argument "dur".')
+    if 'SR' not in argnames:
+        raise ValueError('Function invalid, must have the argument "SR".')
+
+    argnames.remove('SR')
+    argnames.remove('dur')
+
+    if not set(argnames) == set(args_dict.keys()):
+        raise ValueError('Invalid args_dict. args_dict specifies '
+                         f'{set(args_dict.keys())}, but the function '
+                         f'expects {set(argnames)}.')
+
+
+class Segment:
+    """
+    The smallest broadbean particle
+    """
+
+    def __init__(self, name: str, function: Callable,
+                 args_dict: ArgsDict,
+                 duration: Optional[float]=None) -> None:
```

Comment:
I thought we agreed on something slightly different here. But let's discuss:
The problem I see with this implementation is that it does not enable you to make duration a symbol.
One could change the typing and add a line to the symbols definition. I would rather make duration a part of the args_dict. This makes args dict less homogeneous, because a certain arg is required but to my mind this is cleaner than repeating code in a couple of places.

Reply:
Yeah, I wasn't quite sure what you wanted for the duration; I think it will only become apparent with the grouping of segment (like: make this entire group have duration XX s).